renderc/test/render_test.cc

```diff
@@ -0,0 +1,297 @@
+  }
+
+  ASSERT_EQ(index.simple_span.size(), 0);
+
+  {
+    auto span = index.simple_span.alloc();
+    ASSERT_TRUE(span.valid());
+
+    ASSERT_EQ(index.simple_span.size(), 1);
+
+    span.put();
+    ASSERT_FALSE(span.valid());
+
+    ASSERT_EQ(index.simple_span.size(), 0);
+  }
+}
+
+TEST(RenderTest, Handle)
+{
+  static constexpr u32 the_number = 42;
+
+  test::app1::Index index;
+
+  auto auto_handle = index.simple_span.alloc();
+  ASSERT_TRUE(auto_handle.valid());
+
+  // Set the integer field to some important number.
+  auto_handle.modify().number(the_number);
+  ASSERT_EQ(auto_handle.number(), the_number);
+
+  // Convert auto-handle to handle.
+  auto handle = auto_handle.to_handle();
+  ASSERT_TRUE(handle.valid());
+
+  // Auto-handle is released.
+  ASSERT_FALSE(auto_handle.valid());
+  ASSERT_EQ(index.simple_span.size(), 1);
+
+  // Check that it's the same span.
+  ASSERT_EQ(handle.access(index).number(), the_number);
+
+  handle.put(index);
+  ASSERT_FALSE(handle.valid());
+
+  ASSERT_EQ(index.simple_span.size(), 0);
+}
+
+TEST(RenderTest, IndexedSpan)
+{
+  static constexpr u32 key = 42;
+
+  test::app1::Index index;
+
+  {
+    auto ahandle = index.indexed_span.by_key(key);
+    ASSERT_TRUE(ahandle.valid());
+    ASSERT_EQ(ahandle.number(), key);
+
+    ASSERT_EQ(index.indexed_span.size(), 1);
+
+    {
+      auto another = index.indexed_span.by_key(key);
+      ASSERT_TRUE(another.valid());
+
+      // Still only one span is allocated.
+      ASSERT_EQ(index.indexed_span.size(), 1);
+
+      // It's the same span.
+      ASSERT_EQ(ahandle.loc(), another.loc());
+    }
+
+    {
+      auto different = index.indexed_span.by_key(key + 1);
+      ASSERT_TRUE(different.valid());
+
+      // Additional span has been allocated.
+      ASSERT_EQ(index.indexed_span.size(), 2);
+
+      // It's not the same span.
+      ASSERT_NE(ahandle.loc(), different.loc());
+    }
+  }
+
+  ASSERT_EQ(index.indexed_span.size(), 0);
+}
+
+TEST(RenderTest, MetricStore)
+{
+  using metrics_visitor_t =
+      std::function<void(u64, test::app1::weak_refs::metrics_span, test::metrics::some_metrics const &, u64)>;
+
+  static constexpr u64 timeslot_duration = 1'000'000'000;
+  u64 time_now = 1;
+
+  test::app1::Index index;
+
+  {
+    auto span = index.metrics_span.alloc();
+    ASSERT_TRUE(span.valid());
+
+    ASSERT_EQ(index.metrics_span.size(), 1);
+    ASSERT_EQ(span.refcount(), 1);
+
+    test::metrics::some_metrics_point metrics{
+        .active = 55,
+        .total = 100,
+    };
+    span.metrics_update(time_now, metrics);
+
+    ASSERT_EQ(index.metrics_span.size(), 1);
+
+    // Metric store is keeping a reference to this span.
+    ASSERT_EQ(span.refcount(), 2);
+  }
+
+  // Metric store is keeping the span allocated.
+  ASSERT_EQ(index.metrics_span.size(), 1);
+
+  // Advance the current time.
+  time_now += 2 * timeslot_duration;
+
+  // Metrics slot should be ready.
+  ASSERT_TRUE(index.metrics_span.metrics_ready(time_now));
```

Comment:
this is amazing, so getting a bit greedy.. can this also `ASSERT_FALSE(metrics_ready())` when the `time_now` is too early? can help catch bugs in `metrics_ready`